packages/instantsearch.js/src/lib/server.ts

```diff
@@ -37,17 +55,28 @@ export function waitForResults(search: InstantSearch): Promise<void> {
 /**
  * Walks the InstantSearch root index to construct the initial results.
  */
-export function getInitialResults(rootIndex: IndexWidget): InitialResults {
+export function getInitialResults(
+  rootIndex: IndexWidget,
+  /**
+   * Search parameters sent to the search client, usually
+   * returned by `waitForResults()`.
+   */
+  requestParams?: RequestParams
+): InitialResults {
   const initialResults: InitialResults = {};
 
   walkIndex(rootIndex, (widget) => {
     const searchResults = widget.getResults();
     if (searchResults) {
-      initialResults[widget.getIndexId()] = {
+      const indexId = widget.getIndexId();
+      initialResults[indexId] = {
         // We convert the Helper state to a plain object to pass parsable data
         // structures from server to client.
         state: { ...searchResults._state },
         results: searchResults._rawResults,
+        requestParams: requestParams?.find(
+          ({ indexName }) => indexName === indexId
```

Comment:
this should use the widget's index name, not index id

Reply:
Ah I see. I'll update the sample code to use multiple of the same index with different ids and check if I can match request + results in the hydration step.

Reply:
I simplified the implementation as extracted request params and initial results are ordered identically.